MoreLinq.Test/BacksertTest.cs

```diff
@@ -0,0 +1,105 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#endregion
+
+namespace MoreLinq.Test
+{
+    using System;
+    using NUnit.Framework;
+
+    [TestFixture]
+    public class BacksertTest
+    {
+        [Test]
+        public void BacksertWithNegativeIndex()
+        {
+            AssertThrowsArgument.OutOfRangeException("index", () =>
+                 Enumerable.Range(1, 10).Backsert(new[] { 97, 98, 99 }, -1));
+        }
+
+        [Test]
+        public void BacksertWithIndexGreaterThanSourceLength()
+        {
+            const int count = 5;
+            var seq1 = Enumerable.Range(0, count);
+            var seq2 = new[] { 97, 98, 99 };
+
+            using (var test1 = seq1.AsTestingSequence())
+            using (var test2 = seq2.AsTestingSequence())
+            {
+                var result = test1.Backsert(test2, count + 1);
+
+                Assert.Throws<ArgumentOutOfRangeException>(() => result.ElementAt(0));
+            }
+        }
+
+        [Test]
+        public void BacksertWithIndexEqualsSourceLength()
+        {
+            const int count = 5;
+            var seq1 = Enumerable.Range(1, count);
+            var seq2 = new[] { 9 };
+
+            using (var test1 = seq1.AsTestingSequence())
+            using (var test2 = seq2.AsTestingSequence())
+            {
+                var result = test1.Backsert(test2, count);
+                var expectations = seq2.Concat(seq1);
+
+                Assert.That(result, Is.EqualTo(expectations));
+            }
+        }
+
+        [Test]
+        public void BacksertWithIndexZero()
+        {
+            var seq1 = Enumerable.Range(1, 5);
+            var seq2 = new[] { 9 };
+
+            using (var test1 = seq1.AsTestingSequence())
+            using (var test2 = seq2.AsTestingSequence())
+            {
+                var result = test1.Backsert(test2, 0);
+                var expectations = seq1.Concat(seq2);
+
+                Assert.That(result, Is.EqualTo(expectations));
+            }
+        }
+
+        [TestCase(3, 1)]
+        [TestCase(3, 2)]
+        public void Backsert(int count, int index)
+        {
+            var seq1 = Enumerable.Range(1, count);
+            var seq2 = new[] { 97, 98, 99 };
+
+            using (var test1 = seq1.AsTestingSequence())
+            using (var test2 = seq2.AsTestingSequence())
+            {
+                var result = test1.Backsert(test2, index);
+                var expectations = seq1.SkipLast(index).Concat(seq2).Concat(seq1.TakeLast(index));
```

Comment:
I think this test (and others) rely too much on algorithmic expression of expectations that's hard to follow. It makes reviewing the test hard, that it's testing the right thing in the first place without doing the mental gymnastics. In other words, a literal expression is much clearer as to the inputs and what's going to happen:

```c#
[TestCase(new[] { 1, 2, 3 }, 1, new[] { 97, 98, 99 }, ExpectedResult = new[] { 1, 2, 97, 98, 99, 3 })]
[TestCase(new[] { 1, 2, 3 }, 2, new[] { 97, 98, 99 }, ExpectedResult = new[] { 1, 97, 98, 99, 2, 3 })]
public IEnumerable<int> Backsert(int[] seq1, int index, int[] seq2)
{
    using (var test1 = seq1.AsTestingSequence())
    using (var test2 = seq2.AsTestingSequence())
    {
        foreach (var item in test1.Backsert(test2, index))
            yield return item;
    }
}
```

Reply:
I removed all "mental gymnastic"

Reply:
> I removed all "mental gymnastic"

I'm afraid it's still there unless I'm missing something? For example:

```c#
var expectations = seq1.SkipLast(index).Concat(seq2).Concat(seq1.TakeLast(index));
```

Why not use `ExpectedResult` on `[TestCase]` instead? If you do, you can fold `BacksertWithIndexEqualsSourceLength` and `BacksertWithIndexZero` into the `Backset` test.

Reply:
I know, but I prefer keep different tests for each scenario. I think explict tests improves clarity of what is being tested.

Reply:
>  I think explict tests improves clarity of what is being tested.

It's not, unfortunately, which is why I am asking for the changes. I am still doing “mental gymnastics” to do the review so it's not helping. That means when I have 5 minutes and I try to make progress on a review for a _simple operator_ and 5 minutes become 15 and then I have to abort; that translates into the review staying open longer.

We have 3 tests with duplicate code that needs maintenance over time. The only explicit or unique part about each test is its name. If you wish to make a note of that, then I'd comment on top of the `[TestCase]` or even use its `Description` property.

Reply:
I see. So basically "less is more". Okay. Adjust done.